File: Core Data/Core Data Sample Codes/iPhoneCoreDataRecipes/iPhoneCoreDataRecipes/Classes/ImperialPickerController.m
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 Controller to managed a picker view displaying imperial weights.
 */

#import "ImperialPickerController.h"

@interface ImperialPickerController () <UIPickerViewDataSource, UIPickerViewDelegate>

@property (nonatomic, strong) IBOutlet UIPickerView *pickerView;
@property (nonatomic, strong) IBOutlet UILabel *label;

@end


#pragma mark -

@implementation ImperialPickerController

// Identifiers and widths for the various components.
#define POUNDS_COMPONENT 0
#define POUNDS_COMPONENT_WIDTH 110
#define POUNDS_LABEL_WIDTH 60

#define OUNCES_COMPONENT 1
#define OUNCES_COMPONENT_WIDTH 106
#define OUNCES_LABEL_WIDTH 56


// Identifies for component views.
#define VIEW_TAG 41
#define SUB_LABEL_TAG 42
#define LABEL_TAG 43

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
	
	return 2;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
	
	// Number of rows depends on the currently-selected unit and the component.
    if (component == POUNDS_COMPONENT) {
		return 29;
	}
	// OUNCES_LABEL_COMPONENT
	return 16;
}

- (UIView *)labelCellWithWidth:(CGFloat)width rightOffset:(CGFloat)offset {
	
	// Create a new view that contains a label offset from the right.
	CGRect frame = CGRectMake(0.0, 0.0, width, 32.0);
	UIView *view = [[UIView alloc] initWithFrame:frame];
	view.tag = VIEW_TAG;
	
	frame.size.width = width - offset;
	UILabel *subLabel = [[UILabel alloc] initWithFrame:frame];
	subLabel.textAlignment = NSTextAlignmentRight;
	subLabel.backgroundColor = [UIColor clearColor];
	subLabel.font = [UIFont systemFontOfSize:24.0];
	subLabel.userInteractionEnabled = NO;
	
	subLabel.tag = SUB_LABEL_TAG;
	
	[view addSubview:subLabel];
	return view;
}

- (UIView *)pickerView:(UIPickerView *)pickerView viewForRow:(NSInteger)row forComponent:(NSInteger)component reusingView:(UIView *)view {
	
	UIView *returnView = nil;
	
	// Reuse the label if possible, otherwise create and configure a new one.
	if ((view.tag == VIEW_TAG) || (view.tag == LABEL_TAG)) {
		returnView = view;
	}
	else {
        if (component == POUNDS_COMPONENT) {
            returnView = [self labelCellWithWidth:POUNDS_COMPONENT_WIDTH rightOffset:POUNDS_LABEL_WIDTH];
        }
        else {
            returnView = [self labelCellWithWidth:OUNCES_COMPONENT_WIDTH rightOffset:OUNCES_LABEL_WIDTH];
        }
	}
	
	// The text shown in the component is just the number of the component.
	NSString *text = [NSString stringWithFormat:@"%ld", (long)row];
	
	// Where to set the text in depends on what sort of view it is.
	UILabel *theLabel = nil;
	if (returnView.tag == VIEW_TAG) {
		theLabel = (UILabel *)[returnView viewWithTag:SUB_LABEL_TAG];
	}
	else {
		theLabel = (UILabel *)returnView;
	}
    
	theLabel.text = text;
	return returnView;
}

- (CGFloat)pickerView:(UIPickerView *)pickerView widthForComponent:(NSInteger)component {
	
	if (component == POUNDS_COMPONENT) {
		return POUNDS_COMPONENT_WIDTH;
	}
	// OUNCES_COMPONENT
	return OUNCES_COMPONENT_WIDTH;
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    
	// If the user chooses a new row, update the label accordingly.
	[self updateLabel];
}

- (void)updateLabel {

    /*
     If the user has entered imperial units, find the number of pounds and ounces and convert that to kilograms and grams.
     Don't display 0 kg.
     */
    NSInteger ounces = [self.pickerView selectedRowInComponent:OUNCES_COMPONENT];
    ounces += [self.pickerView selectedRowInComponent:POUNDS_COMPONENT] * 16;
    
    float grams = ounces * 28.349;
    if (grams > 1000.0) {
        NSInteger kg = grams / 1000;
        grams -= kg *1000;
        self.label.text = [NSString stringWithFormat:@"%ld kg  %1.0f g", (long)kg, grams];
    }
	else {
        self.label.text = [NSString stringWithFormat:@"%1.0f g", grams];
    }
}
	
@end

File: Core Data/Core Data Sample Codes/iPhoneCoreDataRecipes/Recipes_Swift_4_Advanced/Recipes_Swift/Converter View Controllers/ImperialPickerController.swift
/*
  ImperialPickerController.swift
  Recipes_Swift

  Created by Seyed Samad Gholamzadeh on 2/28/18.
  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.

Abstract:
Controller to managed a picker view displaying imperial weights.
*/


import UIKit

class ImperialPickerController: NSObject, UIPickerViewDataSource, UIPickerViewDelegate {

	@IBOutlet weak var pickerView: UIPickerView!
	@IBOutlet weak var label: UILabel!

	
	// Identifiers and widths for the various components.
	enum Pounds: Int {
		case component = 0, componentWidth = 110, labelWidth = 60
	}
	
	enum Ounces: Int {
		case component = 1, componentWidth = 106, labelWidth = 56
	}
	
	// Identifies for component views.
	enum Tag: Int {
		case view = 41, subLabel, label
	}

	func numberOfComponents(in pickerView: UIPickerView) -> Int {
		return 2
	}
	
	func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
		
		// Number of rows depends on the currently-selected unit and the component.
		if (component == Pounds.component.rawValue) {
			return 29;
		}
		// OUNCES_LABEL_COMPONENT
		return 16;
	}
	
	func labelCellWith( width: CGFloat, rightOffset offset: CGFloat) -> UIView {
		
		// Create a new view that contains a label offset from the right.
		var frame = CGRect(origin: .zero, size: CGSize(width: width, height: 32.0))
		let view = UIView(frame: frame)
		view.tag = Tag.view.rawValue
		
		frame.size.width = width - offset
		let sublabel = UILabel(frame: frame)
		sublabel.textAlignment = .right
		sublabel.backgroundColor = UIColor.clear
		sublabel.font = UIFont.systemFont(ofSize: 24.0)
		
		sublabel.tag = Tag.subLabel.rawValue

		view.addSubview(sublabel)
		return view
	}
	
	func pickerView(_ pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusing view: UIView?) -> UIView {
		
		var returnView: UIView? = nil

		// Reuse the label if possible, otherwise create and configure a new one.
		if view?.tag == Tag.view.rawValue || view?.tag == Tag.label.rawValue {
			returnView = view
		}
		else {
			if component == Pounds.component.rawValue {
				returnView = self.labelCellWith(width: CGFloat(Pounds.componentWidth.rawValue), rightOffset: CGFloat(Pounds.labelWidth.rawValue))
			}
			else {
				returnView = self.labelCellWith(width: CGFloat(Ounces.componentWidth.rawValue), rightOffset: CGFloat(Ounces.labelWidth.rawValue))
			}
		}
		
		// The text shown in the component is just the number of the component.
		let text = "\(row)"
		
		// Where to set the text in depends on what sort of view it is.
		var theLabel: UILabel? = nil
		if returnView?.tag == Tag.view.rawValue {
			theLabel = returnView?.viewWithTag(Tag.subLabel.rawValue) as? UILabel
		}
		else {
			theLabel = returnView as? UILabel
		}
		
		theLabel?.text = text
		return returnView!
	}
	
	func pickerView(_ pickerView: UIPickerView, widthForComponent component: Int) -> CGFloat {
		
		if component == Pounds.component.rawValue {
			return CGFloat(Pounds.componentWidth.rawValue)
		}
		// OUNCES_COMPONENT
		return CGFloat(Ounces.componentWidth.rawValue)
	}
	
	func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
		
		// If the user chooses a new row, update the label accordingly.
		self.updateLabel()
	}
	
	func updateLabel() {
		
		/*
		If the user has entered imperial units, find the number of pounds and ounces and convert that to kilograms and grams.
		Don't display 0 kg.
		*/
		var ounces = self.pickerView.selectedRow(inComponent: Ounces.component.rawValue)
		ounces += self.pickerView.selectedRow(inComponent: Pounds.component.rawValue*16)
		
		var grams: Double = Double(ounces)*28.349
		if grams > 1000.0 {
			let kg = Int(grams/1000)
			grams -= Double(kg*1000)
			self.label.text = "\(kg) kg \(grams) g"
		}
		else {
			self.label.text = "\(grams) g"
		}
	}
}
